sentence_transformers/util.py

```diff
@@ -702,7 +714,19 @@ def mine_hard_negatives(
 
     # Scores is a [num_queries, range_max + 1] tensor, where we set the values to -inf to disqualify the corresponding
     # text as a negative candidate. Here we disqualify the positive pair
-    positive_indices = indices == torch.arange(len(queries), device=indices.device).unsqueeze(-1)
+
+    # positive_indices = indices == torch.arange(len(queries), device=indices.device).unsqueeze(-1)
+    # scores[positive_indices] = -float("inf")
+    positive_indices = []
+    for positive in positives:
+        if positive in corpus:
+            index = corpus.index(positive)
+            positive_indices.append(index)
+        else:
+            positive_indices.append(-1)
+    positive_indices = torch.Tensor(positive_indices).to(torch.int32).unsqueeze(-1)
+    positive_indices = indices == positive_indices
```

Comment:
I did a bit of tests with this, and I ran into a bit of an issue (that already existed, too!):
The same text can occur multiple times in `corpus`/`positives`. For example:
```python
pairs_dataset = Dataset.from_dict({
    "query": ["What is the capital of France?", "What is the capital of Germany?", "What is the capital of Italy?", "What is Paris Hilton's first name?"],
    "positive": ["Paris", "Berlin", "Rome", "Paris"],
})
```
If this is your input `dataset`, and you don't specify a separate `corpus`, then your scores before the `topk` might be:
```python
tensor([[0.6274, 0.3247, 0.3938, 0.6274],
        [0.3650, 0.6055, 0.3545, 0.3650],
        [0.3440, 0.3384, 0.5845, 0.3440],
        [0.5039, 0.1982, 0.2269, 0.5039]], dtype=torch.float16)
```
This makes sense: the 1st and 4th column are identical because the 1st and 4th corpus element is identical. After topk, we get:
```python
tensor([[0.6274, 0.6274, 0.3938, 0.3247],
        [0.6055, 0.3650, 0.3650, 0.3545],
        [0.5845, 0.3440, 0.3440, 0.3384],
        [0.5039, 0.5039, 0.2269, 0.1982]], dtype=torch.float16)
```
And then the `positive_indices` become:
```
tensor([[ True, False, False, False],
        [ True, False, False, False],
        [ True, False, False, False],
        [ True, False, False, False]])
```
because this only finds the first occurrence, not all occurrences. We have this same problem with the new
```python
    positive_indices = []
    for positive in positives:
        if positive in corpus:
            index = corpus.index(positive)
            positive_indices.append(index)
        else:
            positive_indices.append(-1)
    positive_indices = torch.Tensor(positive_indices).to(torch.int32).unsqueeze(-1)
    positive_indices = indices == positive_indices
```
as well as the old
```python
    positive_indices = indices == torch.arange(len(queries), device=indices.device).unsqueeze(-1)
```

I'll try to think of something to counteract this.

- Tom Aarsen

Reply:
@tomaarsen 
Oh nice, great catch! 

I agree that the easiest is to remove duplicates from the `corpus`. I think assuming/forcing the corpus to contain only non-duplicates is also a reasonable assumption.  As an alternative, we could try to find all occurences of a positive in the corpus and thus keep the corpus intact, but that might be computationally less efficient? But in general, I guess it makes sense that during retrieval you retrieve different documents and no duplicates (so advocating the remove duplicate option for corpus).

Regarding removing duplicates of the positives, in some parts we require that positives and queries have the same size (i.e. calculate similarity_pairwise), so it is good to keep them intact?

Reply:
>  As an alternative, we could try to find all occurences of a positive in the corpus and thus keep the corpus intact, but that might be computationally less efficient?

Beyond being computationally less efficient, I now realise that keeping duplicates can hurt the quality of the mined negatives. In particular, the script searches for the `range_max + 1` most similar embeddings, and then does filtering based on e.g. `margin`, `max_score`, etc. Lastly, we take the `num_negatives` negatives from the `range_max + 1` that were not filtered away. If we have duplicates, then it'll occur more often that there are not `num_negatives` non-filtered negative candidates. With other words: we'll be able to find (slightly) more useful negatives if we remove duplicates.

> Regarding removing duplicates of the positives, in some parts we require that positives and queries have the same size (i.e. calculate similarity_pairwise), so it is good to keep them intact?

This is indeed a tricky part. One idea that I had was to remove duplicates in `corpus`, but not in `positives`. Of course, this is a bit tricky to do efficiently, but I think I can make this work. Should I get started on this?

Reply:
> With other words: we'll be able to find (slightly) more useful negatives if we remove duplicates.
Yes, I totally agree!

> This is indeed a tricky part. One idea that I had was to remove duplicates in corpus, but not in positives. Of course, this is a bit tricky to do efficiently, but I think I can make this work. Should I get started on this?
Yes, for me that also feels the way to go now! :) Great to discuss with you Tom!

Reply:
I believe I solved it via 46ae230b6938b74ab44391a78170022b72cf53ab. The patch is smaller than expected, but I believe it works well. In essence, I:
1. deduplicate the corpus naively (via list & set)
2. create a mapping from positive indices to corpus indices
3. use the aforementioned mapping to get the full positive embeddings from the deduplicates corpus embeddings
4. use the aforementioned mapping to recognize when a candidate negative from the corpus is equal to a positive, so I can filter those positives away

I think this works pretty nicely: less embeddings required too

Reply:
This also allows us to simplify the rescoring filtering for positive pairs: aba1fb77d8cc80a152d2d4a62703c47052970d83